### PR TITLE
Support Pselect6 Syscall

### PR DESCRIPTION
**Pull request by ClawSeven**

### CODE REVIEW

**Comment by ClawSeven** [comment]
Reference: https://linux.die.net/man/2/pselect6

**Comment by ClawSeven** [comment]
TODO: replace alarm syscall with kill syscall and user-defined signals.

File: src/libos/src/interrupt/mod.rs [hunk 1]
```diff
@@ -38,11 +38,15 @@ pub fn do_handle_interrupt(
 /// Broadcast interrupts to threads by sending POSIX signals.
 pub fn broadcast_interrupts() -> Result<usize> {
     let should_interrupt_thread = |thread: &&ThreadRef| -> bool {
-        // TODO: check Thread::sig_mask to reduce false positives
-        thread.process().is_forced_to_exit()
+        let should_interrupt = thread.process().is_forced_to_exit()
             || thread.is_forced_to_stop()
-            || !thread.sig_queues().read().unwrap().empty()
-            || !thread.process().sig_queues().read().unwrap().empty()
+            || !thread.process().sig_queues().read().unwrap().empty();
```
Comment (volcano0dr) [inline src/libos/src/interrupt/mod.rs:43]: I think sigmask should be checked for process signal queue.

File: src/libos/src/signal/do_sigtimedwait.rs [hunk 2]
```diff
@@ -129,6 +162,20 @@ impl Drop for PendingSigWaiter {
     }
 }
 
+fn has_interest_signal(interest: &SigSet, thread: &ThreadRef, process: &ProcessRef) -> bool {
+    let blocked = !*interest;
+    process
+        .sig_queues()
+        .write()
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply src/libos/src/signal/do_sigtimedwait.rs]: Done

File: src/libos/src/signal/sig_queues.rs [hunk 3]
```diff
@@ -31,6 +38,41 @@ impl SigQueues {
         self.count == 0
     }
 
+    pub fn has_valid_signal(&self, blocked: &SigSet) -> bool {
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply src/libos/src/signal/sig_queues.rs]: Done

File: src/libos/src/net/syscalls.rs [hunk 4]
```diff
@@ -723,6 +723,112 @@ pub fn do_select(
...
+    if !sig_data.is_null() {
+        from_user::check_ptr(sig_data)?;
+        let user_sig_data = unsafe { &*(sig_data) };
+
+        is_set_sig = !user_sig_data.ss.is_null();
+        if is_set_sig {
+            if user_sig_data.ss_len != std::mem::size_of::<sigset_t>() {
+                return_errno!(EINVAL, "unexpected sigset size");
+            }
+            let mut curr_mask = thread.sig_mask().write().unwrap();
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply src/libos/src/net/syscalls.rs]: Done

File: src/libos/src/signal/do_sigsuspend.rs [hunk 5]
```diff
@@ -0,0 +1,46 @@
+use super::constants::*;
+use super::do_sigprocmask::do_rt_sigprocmask;
+use super::do_sigtimedwait::PendingSigWaiter;
+use super::{sigset_t, MaskOp, SigNum, SigSet, Signal};
+use crate::prelude::*;
+
+pub fn do_sigsuspend(mask: &sigset_t) -> Result<()> {
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply src/libos/src/signal/do_sigsuspend.rs]: Done

File: test/pselect/main.c [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    sigemptyset(&sigmask);
+    sigaddset(&sigmask, SIGUSR1);
+
+    // Access pthread id
+    pthread_t main_thread_id = pthread_self();
+
+    // Spawn new thread for sending signal when call pselect syscall
+    pthread_t signal_thread;
+    if (pthread_create(&signal_thread, NULL, send_signal, &main_thread_id) != 0) {
+        perror("pthread_create");
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply test/pselect/main.c]: Done

File: test/pselect/main.c [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        return 1;
+    }
+
+    int timer_fd = timerfd_create(CLOCK_REALTIME, 0);
+    if (timer_fd == -1) {
+        perror("timerfd_create");
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply test/pselect/main.c]: Done

File: test/pselect/main.c [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        return 1;
+    }
+
+    struct itimerspec timerValue;
+    timerValue.it_value.tv_sec = 2;
+    timerValue.it_value.tv_nsec = 0;
+    timerValue.it_interval.tv_sec = 0;
+    timerValue.it_interval.tv_nsec = 0;
+    if (timerfd_settime(timer_fd, 0, &timerValue, NULL) == -1) {
+        perror("timerfd_settime");
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply test/pselect/main.c]: Done

File: test/pselect/main.c [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        return 1;
+    }
+
+    fd_set readfds;
+    FD_ZERO(&readfds);
+    FD_SET(timer_fd, &readfds);
+
+    int ready = pselect(timer_fd + 1, &readfds, NULL, NULL, NULL, &sigmask);
+
+    if (ready == -1) {
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply test/pselect/main.c]: Done

File: src/libos/src/interrupt/mod.rs (hunk 1 above)
Comment (ClawSeven) [reply src/libos/src/interrupt/mod.rs]: Done

File: src/libos/src/signal/do_sigsuspend.rs [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+use crate::prelude::*;
+
+pub fn do_sigsuspend(mask: &SigSet) -> Result<()> {
+    debug!("do_sigsuspend: mask: {:?}", mask);
+
+    let thread = current!();
+    let process = thread.process().clone();
+
+    // Set signal mask
+    let mut curr_mask = thread.sig_mask().write().unwrap();
```
Comment (volcano0dr) [inline src/libos/src/signal/do_sigsuspend.rs:14]: move to 24 line

File: src/libos/src/signal/sig_queues.rs [hunk 11]
```diff
@@ -130,17 +131,29 @@ impl SigQueues {
         &self.notifier
     }
 
-    pub fn pending(&self) -> SigSet {
+    pub fn pending(&self, blocked: Option<&SigSet>) -> SigSet {
         let mut pending_sigs = SigSet::new_empty();
         for signum in MIN_STD_SIG_NUM..=MAX_STD_SIG_NUM {
             let signum = unsafe { SigNum::from_u8_unchecked(signum) };
+            // Filter blocked signals
+            if let Some(blocked) = blocked {
```
Comment (volcano0dr) [inline src/libos/src/signal/sig_queues.rs:139]:
```
let blocked = blocked.cloned().wrap_or(SigSet::default());
for signum in MIN_STD_SIG_NUM..=MAX_STD_SIG_NUM {
    ......
    if blocked.contains(signum) {
        continue;
    }
    ......
}

```

File: src/libos/src/signal/do_sigpending.rs [hunk 12]
```diff
@@ -6,8 +6,9 @@ pub fn do_sigpending() -> Result<SigSet> {
 
     let thread = current!();
     let process = thread.process();
-    let pending = (thread.sig_queues().read().unwrap().pending()
-        | process.sig_queues().read().unwrap().pending())
-        & *thread.sig_mask().read().unwrap();
+
+    let blocked = thread.sig_mask().read().unwrap();
```
Comment (volcano0dr) [inline src/libos/src/signal/do_sigpending.rs:10]: `let blocked = *thread.sig_mask().read().unwrap();`

File: src/libos/src/interrupt/mod.rs [hunk 13]
```diff
@@ -38,11 +38,16 @@ pub fn do_handle_interrupt(
 /// Broadcast interrupts to threads by sending POSIX signals.
 pub fn broadcast_interrupts() -> Result<usize> {
     let should_interrupt_thread = |thread: &&ThreadRef| -> bool {
-        // TODO: check Thread::sig_mask to reduce false positives
-        thread.process().is_forced_to_exit()
-            || thread.is_forced_to_stop()
-            || !thread.sig_queues().read().unwrap().empty()
-            || !thread.process().sig_queues().read().unwrap().empty()
+        let should_interrupt = thread.process().is_forced_to_exit() || thread.is_forced_to_stop();
```
Comment (volcano0dr) [inline src/libos/src/interrupt/mod.rs:41]:
```
   let should_interrupt_thread = |thread: &&ThreadRef| -> bool {
        let *sig_mask = thread.sig_mask().read().unwrap();
        thread.process().is_forced_to_exit()
            || thread.is_forced_to_stop()
            || !thread.sig_queues().read().unwrap().pending(Some(&sig_mask)).empty()
            || !thread.process().sig_queues().read().unwrap().pending(Some(&sig_mask)).empty()
    };
```

File: src/libos/src/signal/mod.rs [hunk 14]
```diff
@@ -34,3 +35,4 @@ mod signals;
 mod syscalls;
 
 pub mod constants;
+pub mod do_sigprocmask;
```
Comment (volcano0dr) [inline src/libos/src/signal/mod.rs:38]: remove `pub` attribute.
modify do_rt_sigprocmask, sigset_t => SigSet

File: src/libos/src/signal/do_sigsuspend.rs [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+    *curr_mask = update_mask;
+    drop(curr_mask);
+
+    // Suspend for interest signal
+    let interest = SigSet::from_c(!update_mask.to_c());
+    let pending_sig_waiter = PendingSigWaiter::new(thread.clone(), process, interest);
+
+    let err = match pending_sig_waiter.suspend() {
+        Ok(_) => {
+            errno!(EINTR, "Wait for EINTR signal successfully")
```
Comment (volcano0dr) [inline src/libos/src/signal/do_sigsuspend.rs:33]: why is `errno!`

Comment (ClawSeven) [reply src/libos/src/signal/do_sigsuspend.rs:33]: `sigsuspend() always returns -1, with [errno](https://man7.org/linux/man-pages/man3/errno.3.html) set to indicate the
       error (normally, EINTR).`
https://man7.org/linux/man-pages/man2/sigsuspend.2.html

File: src/libos/src/signal/do_sigsuspend.rs (hunk 10 above)
Comment (ClawSeven) [reply src/libos/src/signal/do_sigsuspend.rs]: Done

File: src/libos/src/signal/sig_queues.rs (hunk 11 above)
Comment (ClawSeven) [reply src/libos/src/signal/sig_queues.rs]: Done

File: src/libos/src/signal/do_sigpending.rs (hunk 12 above)
Comment (ClawSeven) [reply src/libos/src/signal/do_sigpending.rs]: Done

File: src/libos/src/interrupt/mod.rs (hunk 13 above)
Comment (ClawSeven) [reply src/libos/src/interrupt/mod.rs]: Done

File: src/libos/src/signal/mod.rs (hunk 14 above)
Comment (ClawSeven) [reply src/libos/src/signal/mod.rs]: Done

File: src/libos/src/net/syscalls.rs [hunk 16]
```diff
@@ -723,6 +723,112 @@ pub fn do_select(
     ret
 }
 
+#[derive(Clone, Copy, Debug)]
+#[repr(C)]
+pub struct Pselect6sig {
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply src/libos/src/net/syscalls.rs]: Done

File: src/libos/src/signal/syscalls.rs [hunk 17]
```diff
@@ -89,23 +89,23 @@ pub fn do_rt_sigprocmask(
...
+            Some((op, &set_sig))
         } else {
             None
         }
     };
-    super::do_sigprocmask::do_rt_sigprocmask(op_and_set, old_set)?;
+
+    let old_set = super::do_sigprocmask::do_rt_sigprocmask(op_and_set)?;
+    if !oldset_ptr.is_null() {
+        unsafe { oldset_ptr.write(old_set.to_c()) }
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply src/libos/src/signal/syscalls.rs]: Done

File: src/libos/src/signal/do_sigsuspend.rs [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+        set
+    };
+
+    let mut curr_mask = thread.sig_mask().write().unwrap();
+    let prev_mask = *curr_mask;
+    *curr_mask = update_mask;
+    drop(curr_mask);
+
+    // Suspend for interest signal
+    let interest = SigSet::from_c(!update_mask.to_c());
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply src/libos/src/signal/do_sigsuspend.rs]: Done

**Review by volcano0dr — APPROVED** [review]
LGTM

**Pull request merged**